Code/src/main/java/nl/utwente/group10/haskell/type/ConstT.java

```diff
@@ -62,6 +67,30 @@ public String toHaskellType() {
     }
 
     @Override
+    public ConstT getFresh() {
+        return new ConstT(this.constructor, this.getFreshArgs());
+    }
+
+    /**
+     * Returns an array of fresh arguments. Selectively calls {@code getFresh} on each argument. When the same Type
+     * instance appears multiple times in the arguments no new type is instantiated. Instead, the fresh type is reused.
+     * @return An array of fresh arguments.
+     */
+    protected Type[] getFreshArgs() {
+        List<Type> fresh = new LinkedList<>();
+
+        for (Type arg : this.args) {
+            if (fresh.contains(arg)) {
+                fresh.add(fresh.get(fresh.indexOf(arg)));
```

Comment:
No. I had to look at it for a good five minutes to understand why it works. Conclusion is it *almost* works.

```LinkedList.contains``` and ```LinkedList.indexOf``` use ```Object.equals```. Since a fresh argument is equal to its source (this is a mistake - this is not true for ```VarT```), it would return the fresh equivalent instead of the argument itself. I thought it to be quite clever since a map is not needed.

I'm going to refactor this into a map so it works for ```VarT```s as well.